e2e/e2e_test.go

```diff
@@ -0,0 +1,206 @@
+
+			state, err := sb.Diff()
+			require.NoError(t, err)
+
+			expected := testutil.ReadOrUpdateGoldenFile(t, state, *updateOpt)
+
+			testutil.AssertDiff(t, expected, state)
+		},
+	}
+
+	steps := Steps{create}
+
+	steps.Run(t)
+}
+
+func TestCreatesSecureClusterWithGeneratedCert(t *testing.T) {
+	if testing.Short() {
+		t.Skip("skipping test in short mode.")
+	}
+
+	testLog := zapr.NewLogger(zaptest.NewLogger(t))
+
+	actor.Log = testLog
+
+	sb := testenv.NewDiffingSandbox(t, env)
+	sb.StartManager(t, controller.InitClusterReconcilerWithLogger(testLog))
+
+	b := testutil.NewBuilder("crdb").WithNodeCount(1).WithTLS().WithEmptyDirDataStore()
+
+	create := Step{
+		name: "creates 1-node secure cluster",
+		test: func(t *testing.T) {
+			require.NoError(t, sb.Create(b.Cr()))
+
+			requireClusterToBeReadyEventually(t, sb, b)
+
+			state, err := sb.Diff()
+			require.NoError(t, err)
+
+			expected := testutil.ReadOrUpdateGoldenFile(t, state, *updateOpt)
+
+			testutil.AssertDiff(t, expected, state)
+		},
+	}
+
+	steps := Steps{create}
+
+	steps.Run(t)
+}
+
+func TestUpgradesMinorVersion(t *testing.T) {
+	if testing.Short() {
+		t.Skip("skipping test in short mode.")
+	}
+
+	testLog := zapr.NewLogger(zaptest.NewLogger(t))
+
+	actor.Log = testLog
+
+	sb := testenv.NewDiffingSandbox(t, env)
+	sb.StartManager(t, controller.InitClusterReconcilerWithLogger(testLog))
+
+	builder := testutil.NewBuilder("crdb").WithNodeCount(1).WithTLS().
+		WithImage("cockroachdb/cockroach:v19.2.5").
+		WithPVDataStore("1Gi", "standard" /* default storage class in KIND */)
+
+	steps := Steps{
+		{
+			name: "creates a 1-node secure cluster",
+			test: func(t *testing.T) {
+				require.NoError(t, sb.Create(builder.Cr()))
+
+				requireClusterToBeReadyEventually(t, sb, builder)
+			},
+		},
+		{
+			name: "upgrades the cluster to the next minor version",
+			test: func(t *testing.T) {
+				current := builder.Cr()
+				require.NoError(t, sb.Get(current))
+
+				current.Spec.Image = "cockroachdb/cockroach:v19.2.6"
+				require.NoError(t, sb.Update(current))
+
+				requireClusterToBeReadyEventually(t, sb, builder)
+				requireDbContainersToUseImage(t, sb, current)
+			},
+		},
+	}
+
+	steps.Run(t)
+}
+
+func TestUpgradesMajorVersion(t *testing.T) {
+	if testing.Short() {
+		t.Skip("skipping test in short mode.")
+	}
+
+	testLog := zapr.NewLogger(zaptest.NewLogger(t))
+
+	actor.Log = testLog
+
+	sb := testenv.NewDiffingSandbox(t, env)
+	sb.StartManager(t, controller.InitClusterReconcilerWithLogger(testLog))
+
+	builder := testutil.NewBuilder("crdb").WithNodeCount(1).WithTLS().
+		WithImage("cockroachdb/cockroach:v19.2.6").
+		WithPVDataStore("1Gi", "standard" /* default storage class in KIND */)
+
+	steps := Steps{
+		{
+			name: "creates a 1-node secure cluster",
+			test: func(t *testing.T) {
+				require.NoError(t, sb.Create(builder.Cr()))
+
+				requireClusterToBeReadyEventually(t, sb, builder)
+			},
+		},
+		{
+			name: "upgrades the cluster to the next minor version",
```

Comment:
Can you also test that preserve downgrade option is being set here but not when doing a patch version? I understand that if we fail to do this, the cluster won't become ready (at least that's what I think will happen), but still seems prudent to test.

Reply:
Sorry, did you mean assertions? If yes, I actually think we need to keep e2e tests clean out of implementation details. However, I think we need some kind of fake implementation of interaction with crdb for functional tests where we would have check for that. I or somebody could look into that later.